### PR TITLE
Support parquet datasets with >10k files

### PR DESCRIPTION
**Pull request by lhoestq**
By respecting the 10k files limit per folder when copying them in `refs/convert/parquet` .

To do so I'm using a directory scheme like config/train-part0/0000.parquet up to config/train-part9/9999.parquet when there are >10k files

Close https://github.com/huggingface/datasets-server/issues/2498

cc @guipenedo

### CODE REVIEW

File: services/worker/src/worker/job_runners/config/parquet_and_info.py [hunk 1]
```diff
@@ -136,6 +151,10 @@ def parse_repo_filename(filename: str) -> tuple[str, str]:
     config, split, _ = parts
     if split.startswith(PARTIAL_PREFIX):
         split = split[len(PARTIAL_PREFIX) :]  # noqa: E203
+    if split[-1] in "0123456789":
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply services/worker/src/worker/job_runners/config/parquet_and_info.py:154]: > Though, just check for the last character seems a bit more risky (maybe we have splits like test1, test2) We might have to check the full suffix -part0

that's exactly what it does already

https://github.com/huggingface/datasets-server/blob/d57914c3a15aff1fc2e0ee6bd448e44b73a40342/services/worker/src/worker/job_runners/config/parquet_and_info.py#L154-L157

File: services/worker/tests/job_runners/config/test_parquet_and_info.py [hunk 2]
```diff
@@ -834,3 +835,22 @@ def test_resolve_trust_remote_code() -> None:
...
         is False
     )
+
+
+@pytest.mark.parametrize(
+    "config,split,shard_idx,num_shards,partial,expected_path_in_repo",
+    [
+        ("config", "split", 0, 1, False, "config/split/0000.parquet"),
+        ("config", "split", 0, 1, True, "config/partial-split/0000.parquet"),
+        ("config", "split", 9999, 10_000, False, "config/split/9999.parquet"),
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply services/worker/tests/job_runners/config/test_parquet_and_info.py:845]: 140000 files is not possible, I'll raise an error in this case

File: services/worker/src/worker/job_runners/config/parquet_and_info.py [hunk 3]
```diff
@@ -98,30 +98,45 @@ def repo_file_rfilename_sort_key(repo_file: RepoFile) -> str:
...
         self.shard_idx = shard_idx
+        self.num_shards = num_shards
         self.partial = partial
 
     @property
     def path_in_repo(self) -> str:
         # For paths like "en/partial-train/0000.parquet" in the C4 dataset.
         # Note that "-" is forbidden for split names so it doesn't create directory names collisions.
         partial_prefix = PARTIAL_PREFIX if self.partial else ""
+        # For paths like "en/train-part0/0000.parquet", "en/train-part1/0000.parquet" up to "en/train-part9/9999.parquet".
```
[the remark this answers is not in the document]
Comment (severo) [reply services/worker/src/worker/job_runners/config/parquet_and_info.py:113]: OK. Is it a hard limit? It does not seem so, from https://huggingface.co/docs/hub/repositories-recommendations

They recommend:

> The maximum number of files per folder cannot exceed 10k files per folder. A simple solution is to create a repository structure that uses subdirectories. For example, a repo with 1k folders from 000/ to 999/, each containing at most 1000 files, is already enough.

so... supporting 1M files

**Review by severo — APPROVED** [review]
OK.

I still think we could allow more than 100k, ie 1M. But as we raise an explicit exception, it will be easy to recompute later too...

**Pull request merged**